core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -271,28 +319,46 @@ CompiledScript compileInternal(Script script, Map<String, String> params) {
         CacheKey cacheKey = new CacheKey(scriptEngineService, type == ScriptType.INLINE ? null : name, code, params);
         CompiledScript compiledScript = cache.get(cacheKey);
 
-        if (compiledScript == null) {
-            //Either an un-cached inline script or indexed script
-            //If the script type is inline the name will be the same as the code for identification in exceptions
-            try {
-                // but give the script engine the chance to be better, give it separate name + source code
-                // for the inline case, then its anonymous: null.
-                String actualName = (type == ScriptType.INLINE) ? null : name;
-                compiledScript = new CompiledScript(type, name, lang, scriptEngineService.compile(actualName, code, params));
-            } catch (ScriptException good) {
-                // TODO: remove this try-catch completely, when all script engines have good exceptions!
-                throw good; // its already good
-            } catch (Exception exception) {
-                throw new GeneralScriptException("Failed to compile " + type + " script [" + name + "] using lang [" + lang + "]", exception);
+        if (compiledScript != null) {
+            return compiledScript;
+        }
+
+        // Synchronize so we don't compile scripts many times during multiple shards all compiling a script
+        synchronized (this) {
+            // Retrieve it again in case it has been put by a different thread
+            compiledScript = cache.get(cacheKey);
+
+            if (compiledScript == null) {
+                try {
+                    // Either an un-cached inline script or indexed script
+                    // If the script type is inline the name will be the same as the code for identification in exceptions
+
+                    // but give the script engine the chance to be better, give it separate name + source code
+                    // for the inline case, then its anonymous: null.
+                    String actualName = (type == ScriptType.INLINE) ? null : name;
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("compiling script, type: [{}], lang: [{}], params: [{}]", type, lang, params);
+                    }
+                    compiledScript = new CompiledScript(type, name, lang, scriptEngineService.compile(actualName, code, params));
+                    // Check whether too many compilations have happened, but after compiling, so that scripts that have compilation errors
+                    // don't count towards the limit, since this is not directly mapped to memory it's okay to wait until compilation
+                    // completes before checking the limit
+                    checkCompilationLimit();
```

Comment:
I thought about this some more, and I think even with throwing an exception, breaking the check and decrement into two different pieces doesn't sound as good to me, for two reasons:

- It means the check will no longer be exact, since there will be a period where there are tokens in the bucket but there shouldn't be, because the decrement hasn't happened yet
- This should not be a high-traffic part of the code, scripts should be compiled once which is why we have the cache, the checkCompilationLimit function overhead is going to be very small compared to the actual compilation overhead